### DISCUSSION TITLE
Stereo tool 2.2.3, Stereotool library not found or invalid!

### DISCUSSION THREAD

**Discussion by finnie2006** [savonet/liquidsoap#3088] [User Support]
I am trying to get stereo tool working on the latest liquidsoap version but i keep getting an error about the libary, i am running liquidsoap on a rpi 4,

I was not sure wich libary to use but i guessed libStereoTool_pi2.so is that correct or not?

**Comment by smimram** [comment]
The full path should be specified in the `library_file` argument of the `stereotool` operator:
```
s = stereotool(
  library_file="/path/to/libStereoTool_pi2.so",
  license_key="my_license_key",
  preset="/path/to/preset/file",
  s
)
```

**Comment by finnie2006** [comment]
So if my liquidsoap config is in the same place as the libary file i still  have to do /etc/liquidsoap/libStereoTool_pi2.so instead off just /libStereoTool_pi2.so

**Comment by finnie2006** [comment]
![afbeelding](https://github.com/savonet/liquidsoap/assets/57916807/25b756ab-ab7b-4304-84c8-2c91dde7c317)
![afbeelding](https://github.com/savonet/liquidsoap/assets/57916807/6ba769d2-d45c-4ae4-8f90-5408e7d78d56)

**Comment by toots** [comment]
Hi @finnie2006 . Generally speaking, you need to make sure that the dynamic library matches your architecture. I believe recent rapsberry pi are arm64 so you need a shared library compiled for that architecture.

Quickly glancing at the file, I'm not sure if there is one. I would at least try the different one. Also, I think the `ldd` command line utility can be used against the shared library to see if one is compatible with your architecture.

Lastly, I imagine the difference in path above is not a typo? I just changed the error message, now it should differentiate between file not found and invalid library, this should help clarifying these issues.

**Comment by finnie2006** [comment]
I updated using the deb file https://github.com/savonet/liquidsoap/releases/download/rolling-release-v2.2.x/liquidsoap-2cab684_2.2.0-ubuntu-jammy-1_arm64.deb  

It was modified 8 hours ag, but i still get the combined error message ```message: "Stereotool library not found or invalid!",```

or is the opam installation mode required?

When using ldd i get the following, what indeed seems to indicate that that is not an arm64 binary (i guess)

```
root@server-1:/etc/liquidsoap/libary# ldd libStereoTool_pi2.so
        not a dynamic executable
```        
        
For testing i also ran the ldd command on an actual pi 2 and there it succesfully ran the ldd command which means stereo tool has to compile an serperate arm64 version

**Comment by toots** [comment]
The commit was not yet pushed to rolling release. I just did and it should be there now.

Yes, it looks like a build for `arm64` is missing. I would see with the provider to see if they could add it to their SDK.

**Reply by finnie2006** [reply]
Great, i will wait

**Reply by finnie2006** [reply]
Did you already talk to them?

**Reply by toots** [reply]
No I have not.

**Comment by vitoyucepi** [comment]
Hi @finnie2006,
We can bring the problem to the attention of @hansvanzutphen from Thimeo Audio Technology (the maker of Stereo Tool). At least I can try.

**Reply by hansvanzutphen** [reply]
Oh, you're right, the latest release didn't have a generic plugin for ARM64. You can try last night's beta build:
https://download.thimeo.com/Stereo_Tool_Generic_plugin_1002-078.zip

The file that you need is libStereoTool_arm64.so

Please note: I have not yet properly tested this. I have tested several of the other builds and they are ok, so I don't expect any issues.

**Reply by vitoyucepi** [reply]
Hi @hansvanzutphen,
Is it possible to create a GitHub repo that will serve the binary files?
I think this should be a simple repo with a readme and a bunch of releases. Each release should contain multiple attached release artifacts instead of a single zip archive.
Also the archive uses `\` as a separator instead of `/` for some reason.

**Reply by finnie2006** [reply]
Thanks for including the arm64 libary!

**Reply by finnie2006** [reply]
Tested, working no issues for so far, one question, is it possible to enable the stereo tool web gui or is that not possible using the binary?

**Reply by hansvanzutphen** [reply]
If you load a .sts file with the web interface enabled, it will switch on. Since this is not the command line version I don't think there's any other way t request it to turn the web interface on, and for security reasons I don't want it to be on by default.

**Reply by speniti** [reply]
Hi @hansvanzutphen, I am trying to use the library you have linked, but unfortunately I can't make it work 😔. 
I am on a M1 Mac, and I am trying to compile a Docker image compatible with ARM64. 
Is the `libStereoTool_arm64.so` the correct file to use? 
I have tried also other `.so` files, but I keep getting an `Invalid stereotool library` error.

**Reply by hansvanzutphen** [reply]
Hello @speniti , no, that's the ARM64 Linux version! You'll need the .dylib file, that's a Mac file. It should be a universal binary (eg containing both Intel and M1 code).

**Reply by speniti** [reply]
Thanks @hansvanzutphen, I have tried your solution, but I got the same error 😅. 
I think that since I am running Liquidsoap in a Docker container I should use the Linux one, am I wrong? 

This is the output of the `ldd` command that I run inside the container:

```
liquidsoap@121b4bd3e476:/$ ldd /var/liquidsoap/stereo_tool/lib/libStereoTool_arm64.so 
linux-vdso.so.1 (0x0000ffff8a702000)
librt.so.1 => /lib/aarch64-linux-gnu/librt.so.1 (0x0000ffff89110000)
libX11.so.6 => /lib/aarch64-linux-gnu/libX11.so.6 (0x0000ffff88fb0000)
libbz2.so.1.0 => /lib/aarch64-linux-gnu/libbz2.so.1.0 (0x0000ffff88f80000)
libz.so.1 => /lib/aarch64-linux-gnu/libz.so.1 (0x0000ffff88f40000)
libXpm.so.4 => not found
libm.so.6 => /lib/aarch64-linux-gnu/libm.so.6 (0x0000ffff88ea0000)
libgcc_s.so.1 => /lib/aarch64-linux-gnu/libgcc_s.so.1 (0x0000ffff88e60000)
libpthread.so.0 => /lib/aarch64-linux-gnu/libpthread.so.0 (0x0000ffff88e30000)
libc.so.6 => /lib/aarch64-linux-gnu/libc.so.6 (0x0000ffff88c80000)
/lib/ld-linux-aarch64.so.1 (0x0000ffff8a6c5000)
libxcb.so.1 => /lib/aarch64-linux-gnu/libxcb.so.1 (0x0000ffff88c30000)
libXau.so.6 => /lib/aarch64-linux-gnu/libXau.so.6 (0x0000ffff88c10000)
libXdmcp.so.6 => /lib/aarch64-linux-gnu/libXdmcp.so.6 (0x0000ffff88bf0000)
libbsd.so.0 => /lib/aarch64-linux-gnu/libbsd.so.0 (0x0000ffff88bb0000)
libmd.so.0 => /lib/aarch64-linux-gnu/libmd.so.0 (0x0000ffff88b80000)
```

Seems that the file is correct, but Liquidsoap keeps saying that the library is not valid.
This is the container output:

```
streaming_liquidsoap  | At audio.liq, line 232, char 6-73:
streaming_liquidsoap  | 
streaming_liquidsoap  | Error 14: Uncaught runtime error:
streaming_liquidsoap  | type: invalid,
streaming_liquidsoap  | message: "Invalid stereotool library",
streaming_liquidsoap  | stack: at audio.liq, line 232, char 6-73,
streaming_liquidsoap  | at audio.liq, line 232, char 6-73,
streaming_liquidsoap  | at /var/liquidsoap/radio.liq, line 43 char 9 - line 47 char 1
streaming_liquidsoap  | 
streaming_liquidsoap  | 
streaming_liquidsoap exited with code 1
```

@toots do you have any idea why this happens?

**Reply by speniti** [reply]
Hey guys, I have found the problem 😅, it was my fault I was using the liquidsoap-minimal image instead of the complete one 🤦. Thanks for the help!
